cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -42,13 +42,1440 @@
+// Supports both in-memory and disk-based modes depending on available host memory.
+// In disk mode, the graph is stored in ace_build_dir and dataset is reordered on disk.
+// The returned index is not usable for search. Use the created files for search instead.
+template <typename T, typename IdxT, typename Accessor>
+index<T, IdxT> build_ace(
+  raft::resources const& res,
+  const index_params& params,
+  raft::mdspan<const T, raft::matrix_extent<int64_t>, raft::row_major, Accessor> dataset,
+  size_t num_partitions = 0)
+{
+  common::nvtx::range<common::nvtx::domain::cuvs> function_scope(
+    "cagra::build_ace<%s>(%zu, %zu, %zu)",
+    Accessor::is_managed_type::value ? "managed"
+    : Accessor::is_host_type::value  ? "host"
+                                     : "device",
+    params.intermediate_graph_degree,
+    params.graph_degree,
+    num_partitions);
+
+  uint64_t dataset_size = dataset.extent(0);
+  uint64_t dataset_dim  = dataset.extent(1);
+
+  RAFT_EXPECTS(dataset_size > 0, "ACE: Dataset must not be empty");
+  RAFT_EXPECTS(dataset_dim > 0, "ACE: Dataset dimension must be greater than 0");
+  RAFT_EXPECTS(params.intermediate_graph_degree > 0,
+               "ACE: Intermediate graph degree must be greater than 0");
+  RAFT_EXPECTS(params.graph_degree > 0, "ACE: Graph degree must be greater than 0");
+
+  uint64_t n_partitions = (num_partitions > 0) ? num_partitions : params.ace_npartitions;
+  RAFT_EXPECTS(n_partitions > 0, "ACE: num_partitions must be greater than 0");
+  RAFT_EXPECTS(n_partitions <= dataset_size, "ACE: num_partitions cannot exceed dataset size");
+
+  uint64_t min_vectors_per_partition  = dataset_size / n_partitions;
+  uint64_t min_required_per_partition = 1000;
+  RAFT_EXPECTS(
+    min_vectors_per_partition >= min_required_per_partition,
+    "ACE: Partition size too small. Each partition needs at least %lu vectors, but would have ~%lu",
+    min_required_per_partition,
+    min_vectors_per_partition);
+
+  auto total_start = std::chrono::high_resolution_clock::now();
+  RAFT_LOG_INFO("ACE: Starting partitioned CAGRA build with %zu partitions", n_partitions);
+
+  size_t intermediate_degree = params.intermediate_graph_degree;
+  size_t graph_degree        = params.graph_degree;
+
+  if (intermediate_degree >= dataset_size) {
+    RAFT_LOG_WARN(
+      "Intermediate graph degree cannot be larger than dataset size, reducing it to %lu",
+      dataset_size);
+    intermediate_degree = dataset_size - 1;
+  }
+  if (intermediate_degree < graph_degree) {
+    RAFT_LOG_WARN(
+      "Graph degree (%lu) cannot be larger than intermediate graph degree (%lu), reducing "
+      "graph_degree.",
+      graph_degree,
+      intermediate_degree);
+    graph_degree = intermediate_degree;
+  }
+
+  size_t available_memory = get_free_host_memory<T, IdxT>();
+
+  // Optimistic memory model: focus on largest arrays, assumes all partitions are of equal size
+  // 5 main arrays: partition labels, forward, 2 backward mappings
```

Comment:
How does it add up to five arrays? Why do we have two backward mappings?

Reply:
One for the core dataset and one for the augmented dataset. These are required in the in-memory path to gather the partition's datasets and to map the indices back to the original index space.